### ISSUE TITLE
add `rotate` command?

### ISSUE THREAD

**Issue by motdotla** [dotenvx/dotenvx#478]
should we add a `dotenvx rotate` or `dotenvx rotatekeys` command?

https://stackoverflow.com/questions/79218428/how-to-generate-new-private-and-public-key-pairs-in-dotenvx

would rotate the private/public key pair and re-encrypt everything in one simple command.

**Comment by stephenhuh** [comment]
this seems like a requirement - the CLI as it stands is fairly implicit about this example: [so question](https://stackoverflow.com/questions/79218428/how-to-generate-new-private-and-public-key-pairs-in-dotenvx).

`dotenv-vault` has this as well so would make migration path easier.

also obviously because dotenvx doesn't manage the secret key this seems like a routine command that will be run

**Comment by motdotla** [comment]
Indeed, but there is nuance here:

## Scenario 1: The attacker gets access to both immediately

You are in big trouble. Now your only ally is time - as the attacker has to take the time to decrypt it (and likely they gained access to many other people's at the same time since most of the time these kind of attacks happen on a platform where you would have both stored - AWS, a server, Vercel, etc)

## Scenario 2: The attacker gets access to the private key but not the file

This is the [CircleCI scenario](https://circleci.com/blog/jan-4-2023-incident-report/). 

If you ARE NOT using dotenvx you are in massive trouble - bigger than scenario 1 above - since you don't even have time on your side. The attacker immediately has access to all your real api keys (stripe, twilio, etc) because they were necessarily stored in plain text on CircleCI. Anyone not using dotenvx on platforms like AWS, a server, Vercel, etc still run this risk. It's a massive risk.

If you ARE using dotenvx this is a time when `dotenvx rotate` command like this is very useful. You can create a new encrypted file with a new private key. The attacker's knowledge of your private key is now useless.

## Scenario 3: The attacker gets access to only your encrypted file

This is similarly a situation where `dotenvx rotate` is valuable. The attacker gains access to your encrypted file. Typically then how this goes is a breach is noticed and that is patched up. Then let's say you use `dotenvx rotate` and remove any references to the old private key. Then a couple months later the attacker additionally manages to gain access to your private key. Well that private key is useless against the encrypted file he has.

So in summary, `dotenvx` as it is currently implemented plus a `dotenvx rotate` puts time on your side - valuable time you could then go use to do the slow and tedious work of swapping out all your real keys.

**Comment by jzila** [comment]
> It’s important to communicate that just rotating the dotenvx key provides almost no security by itself. A bulk utility could make it harder to know which real keys have actually been rotated.

I wholeheartedly agree with @kevlened here - re-encrypting existing values to a new key affords no additional security, and may in fact make it harder to perform forensics on whether a secret _value_ was rotated and when.

Dotenvx today already supports multiple keypairs using comma-separated values: `DOTENV_PRIVATE_KEY_FOO=key1,key2`. We should extend this functionality in the following ways:

1. The key ordering becomes explicit - the last key is always the one used to encrypt.
2. The `VALUE='encrypted:<encrypted-value>'` syntax should be extended to use explicit keying: `VALUE='encrypted:<key-fingerprint>:<encrypted-value>'`
  * If the old syntax is present, you attempt decryption with all keys.
  * If the new syntax is used, you _only_ attempt to decrypt with the key with the correct fingerprint. Fail if one doesn't exist.
4. `dotenvx set` and `dotenvx encrypt` are extended to use the latest key.
5. Implement a new `dotenvx key` subcommand for key management. No explicit `rotate` command - you split them to allow phased migration of values to new keypairs.
  * `dotenvx key append` adds a new keypair to the env var in .env.keys and .env.<environment>
  * `dotenvx key revoke` removes a deprecated keypair

These changes maintain backward compatibility while enabling secure rotation functionality. They honor the philosophy of dotenvx to do the right thing by default while enabling advanced functionality only to those who need it. I urge no auto-re-encryption capability to be implemented into the client, so that the canonical use of dotenvx maintains compliance forensics with respect to value rotation.

**Comment by motdotla** [comment]
i need to give the forensics idea more thought. right now dotenvx doesn't concern itself with that - it leaves that git/version-control.

as to wether rotation is selective (choose which keys to rotate) or by default non-selective (rotate all in one fell swoop), certainly the private key should be appended or prepended to DOTENV_PRIVATE_KEY in the .env.keys rather than just replaced. i think the removal we will leave as a manual process.

speaking to some devops people they like the idea of rotating - even though they admit to @kevlened and @jzila's points here that it ultimately doesn't truly solve the underlying problem. (that underlying problem is a very hard one that requires multiple 3rd parties to coordinate - currently far away for the industry so i'm going to ignore that for purpose of this discussion)

the current draft of the [whitepaper](https://dotenvx.com/dotenvx.pdf?v=2) is now commenting on this:

<img width="971" alt="Image" src="https://github.com/user-attachments/assets/6dd99f76-9bba-462a-82d2-88f642332955" />

**Comment by motdotla** [comment]
progress started here: https://github.com/dotenvx/dotenvx/pull/530

**Comment by motdotla** [comment]
this went out in [`1.37.0`](https://github.com/dotenvx/dotenvx/blob/main/CHANGELOG.md#1370)

after we launch pro, we hope/intend to someday also support rotating individual keys to some degree. that's probably at least a year away though. long term it is a major goal for the project though.
